### DISCUSSION TITLE
[Python] What is the purpose of having @odp.base_constructor if it's being assigned to subclass constructors as well?

### DISCUSSION THREAD

**Discussion by anthony-S93** [olimorris/onedarkpro.nvim#168] [Q&A] [answered]
![2023-03-12-130208_536x401_scrot](https://user-images.githubusercontent.com/69449791/224525366-287c24bf-f4e1-4814-973d-4266ea9f0e44.png)

I assume the purpose of `@odp.base_constructor` is so that we can differentiate the highlighting between the base constructor and the constructor of subclasses? 

@olimorris This seems a bit counter-intuitive to me and I was hoping someone can provide an explanation.

**Comment by olimorris** [comment, marked as answer]
Referencing the `after/queries/python/highlights.scm` file:

```schema
((function_definition name: (identifier) @odp.base_constructor)
(#any-of? @odp.base_constructor "__new__" "__init__"))
```

It allows the scheme to reference `__new__` and `__init__` directly. Perhaps it should have been called `odp.initializer`?

**Reply by anthony-S93** [reply]
Hi, @olimorris . Thanks for the explanation. Keeping it as `base_constructor` should be fine, I guess. On a related note, is it possible to do the same for certain built-in classes like `range` or `list` in the context of constructor calls? When we write code like `x = range(10)`, Treesitter treats `range` as `function.builtin`, I suspect simply because it doesn't begin with a capital letter. 

![2023-03-12-185258_486x263_scrot](https://user-images.githubusercontent.com/69449791/224539967-b8d1c4b9-55af-4ed8-b2f5-81df57cf5213.png)

![2023-03-12-185318_472x284_scrot](https://user-images.githubusercontent.com/69449791/224539983-95a71d9a-ef18-4301-a8f3-90d23778ba21.png)

It's not that big of a deal I suppose, though my inner OCD self is screaming at this point. I mean, those two entities are listed as classes according to python's official documentation...

**Reply by olimorris** [reply]
Hmmmmm. Are they not built in functions as per [this](https://docs.python.org/3/library/functions.html#built-in-funds)?

My knowledge of Python isn't great.

**Reply by anthony-S93** [reply]
@olimorris `range` and `list` are actually [built-in classes](https://docs.python.org/3/library/stdtypes.html?highlight=list#list). 
![2023-03-12-195401_405x267_scrot](https://user-images.githubusercontent.com/69449791/224542878-1a631cbe-7dc2-464b-883e-20fcbb13f762.png)

But when I use it in a call expression, it gets parsed as `function.builtin`. 
![2023-03-12-195454_478x381_scrot](https://user-images.githubusercontent.com/69449791/224542910-594d7a3b-2170-4788-8917-6aaef4a7d98c.png)

Other built-in types doesn't have this problem because the class names begin with capital letters, for example `TypeError`

![2023-03-12-195637_639x387_scrot](https://user-images.githubusercontent.com/69449791/224542997-8266fc90-4b7e-492e-991b-7022faad30ca.png)

The same behavior is observed over built-in types that don't begin with capital letters like `int`, `complex`, `float`. 
![2023-03-12-195729_522x230_scrot](https://user-images.githubusercontent.com/69449791/224543047-185032db-bb36-4d88-bbcc-b8938a44bc42.png)

![2023-03-12-195751_569x360_scrot](https://user-images.githubusercontent.com/69449791/224543070-fae5972c-42f7-498c-ac6f-6954a65f1290.png)

Python convention is to write classes in capital letters. Not sure why these built-in types don't begin in caps.

**Reply by olimorris** [reply]
So this should select all of those instances:

```scheme
(assignment right: (call (identifier) @tmp)
(#any-of? @tmp "range" "list"))
```

What would you suppose we name the highlight? Feels like it should be separate from `@odp.base_constructor`.

**Reply by anthony-S93** [reply]
For other Python built-in types (the ones that begin with capital letters), treesitter currently uses `@constructor` for the highlight. 
![2023-03-13-075327_557x286_scrot](https://user-images.githubusercontent.com/69449791/224581838-6cc666bf-47d3-489d-a90b-dc7ae93c4355.png)
Is it possible to follow the same convention? I mean, conceptually, these two calls should be understood as statements that instantiate classes: 

```python
# Both statements create instances of built-in classes
# There's no reason for one to be treated as a constructor call and another to be treated as a function call. 
x = TypeError()
y = list()   
```

**Reply by olimorris** [reply]
I'd be wary of calling it `@constructor` as it would overwrite the Python parser's groups (as I erroneously did in #167). We could call it `@odp.constructor` and then link it to `@constructor`.

**Reply by anthony-S93** [reply]
> I'd be wary of calling it `@constructor` as it would overwrite the Python parser's groups (as I erroneously did in #167). We could call it `@odp.constructor` and then link it to `@constructor`.

@olimorris Sounds good. Will this be implemented?

**Reply by olimorris** [reply]
Have pushed this in the latest commit 👍🏼

**Reply by anthony-S93** [reply]
@olimorris Saw this. Created a pull request to add in the rest of the built-in types.
